[irsdkcpp] Add new port

jglanz opened the pull request:
If this PR adds a new port, please uncomment and fill out this checklist:

- [x] Changes comply with the [maintainer guide](https://github.com/microsoft/vcpkg-docs/blob/main/vcpkg/contributing/maintainer-guide.md).
- [x] The name of the port matches an existing name for this component on https://repology.org/ if possible, and/or is strongly associated with that component on search engines.
- [x] Optional dependencies are resolved in exactly one way. For example, if the component is built with CMake, all `find_package` calls are REQUIRED, are satisfied by `vcpkg.json`'s declared dependencies, or disabled with [CMAKE_DISABLE_FIND_PACKAGE_Xxx](https://cmake.org/cmake/help/latest/variable/CMAKE_DISABLE_FIND_PACKAGE_PackageName.html).
- [x] The versioning scheme in `vcpkg.json` matches what upstream says.
- [x] The license declaration in `vcpkg.json` matches what upstream says.
- [x] The installed as the "copyright" file matches what upstream says.
- [x] The source code of the component installed comes from an authoritative source.
- [x] The generated "usage text" is accurate. See [adding-usage](https://github.com/microsoft/vcpkg-docs/blob/main/vcpkg/examples/adding-usage.md) for context.
- [x] The version database is fixed by rerunning `./vcpkg x-add-version --all` and committing the result.
- [x] Only one version is in the new port's versions file.
- [x] Only one version is added to each modified port's versions file.

jglanz committed:
[irsdkcpp] Adding IRSDKCPP release v1.0.0

jglanz committed:
[irsdkcpp] generated versions and formatting

jglanz commented:
@microsoft-github-policy-service agree

jglanz committed:
[irsdkcpp] Removed static artifacts for dll build & vice versa

jglanz committed:
[irsdkcpp] removed debug/include files in portfile

jglanz committed:
[irsdkcpp] simplified support expression

jglanz committed:
Update version database

LilyWangLL commented on ports/irsdkcpp/portfile.cmake:
```diff
@@ -0,0 +1,29 @@
+vcpkg_check_linkage(ONLY_STATIC_LIBRARY)
```
https://github.com/vrkit-platform/irsdk-cpp/blob/b026e6d01fe511526ff7251276805b1b479cd456/cmake/targets.cmake#L133-L134

```
  ADD_LIBRARY(${sharedTarget} MODULE ${${libSourcesVarName}} ${${libHeadersVarName}} ${exportHeader})
  ADD_LIBRARY(${staticTarget} MODULE ${${libSourcesVarName}} ${${libHeadersVarName}} ${exportHeader})
```

This port has code for building dynamic libraries; why is it specified to build only static libraries here?

jglanz replied:
I've enabled & tested the dynamic library configuration. thanks for the feedback

LilyWangLL commented on ports/irsdkcpp/portfile.cmake:
```diff
@@ -0,0 +1,29 @@
+    OUT_SOURCE_PATH SOURCE_PATH
+    REPO vrkit-platform/irsdk-cpp
+    REF "v${VERSION}"
+    # REF master
```
```suggestion
```

jglanz replied:
I removed the comment, apologies, I was in rapid dev mode

jglanz committed:
[irsdkcpp] enable dynamic library build

jglanz committed:
[irsdkcpp] removed commented out REF

LilyWangLL commented:
Currently, both dynamic and static libraries are being installed, as shown in the image below. Could you add a check for dynamic and static builds to `add_library()`, or remove the shared and static parameters from `add_library()`?

![image](https://github.com/user-attachments/assets/91e27428-34ef-40df-aab1-2fcc6fdde78d)

dg0yt commented on ports/irsdkcpp/portfile.cmake:
```diff
@@ -0,0 +1,25 @@
+vcpkg_cmake_install()
+vcpkg_cmake_config_fixup(CONFIG_PATH lib/cmake/irsdkcpp)
+
+if(VCPKG_LIBRARY_LINKAGE STREQUAL "static")
+  file(REMOVE_RECURSE "${CURRENT_PACKAGES_DIR}/debug/bin" "${CURRENT_PACKAGES_DIR}/bin")
+endif()
```
What is there in static builds: Empty dirs? Executables? Nothing?

jglanz replied:
The original project was part of a larger application, for which it had generated static & dynamic libraries no matter the configuration (BUILD_STATIC/SHARED were not options, but instead both defaulted to ON).

I'm quickly rewriting the CMakeLists tomorrow and implementing the differentiation between SHARED/STATIC

I believe that was/is the underlying question?

thx again, Jon

dg0yt commented:
> ```cmake
> if(VCPKG_LIBRARY_LINKAGE STREQUAL "dynamic")
>   file(REMOVE_RECURSE "${CURRENT_PACKAGES_DIR}/debug/lib" "${CURRENT_PACKAGES_DIR}/lib")
> endif()
> ```

Removing `lib` is always wrong. For Windows dynamic, it holds the import libs. For everything else, it holds the actual libs.

> Hi @dg0yt, just confirming you are suggesting the samething?

Removing `bin` is a suggestion which is currently hard-coded into the tool. But there are only few cases where it is actually the right thing. If there are any executables, they should be handled with  `vcpkg_copy_tools( ... AUTO_CLEAN)`.

jglanz committed:
[irsdkcpp] fully implemented static & dynamic library options

jglanz committed:
[irsdkcpp] Made static build standard

jglanz committed:
[irsdkcpp] Filled in all the missing installation and configuration required to enable CMAKE to produce usable shared & static libraries from the same configuration

jglanz committed:
[irsdkcpp]: removed uwp support

jglanz committed:
Updated usage

jglanz committed:
Merge branch 'microsoft:master' into irsdk-cpp-1.0.0

jglanz committed:
[irsdkcpp]: add cppwinrt for compilation consistency

jglanz committed:
[irsdkcpp]: updated git-tree

LilyWangLL commented on ports/irsdkcpp/usage:
```diff
@@ -0,0 +1,9 @@
+irsdkcpp provides CMake targets:
+
+find_package(irsdkcpp CONFIG REQUIRED)
+
+# STATIC LINKAGE
+target_link_libraries(main PRIVATE irsdkcpp::irsdkcpp_static)
+
+# DYNAMIC LINKAGE
+target_link_libraries(main PRIVATE irsdkcpp::irsdkcpp_shared)
```
```suggestion
  find_package(irsdkcpp CONFIG REQUIRED)
  target_link_libraries(main PRIVATE $<IF:$<TARGET_EXISTS:irsdkcpp::irsdkcpp_shared>,irsdkcpp::irsdkcpp_shared,irsdkcpp::irsdkcpp_static>)
```

LilyWangLL commented on ports/irsdkcpp/portfile.cmake:
```diff
@@ -0,0 +1,29 @@
+vcpkg_fixup_pkgconfig()
+vcpkg_copy_pdbs()
+
+file(REMOVE_RECURSE "${CURRENT_PACKAGES_DIR}/debug/include")
```
```suggestion
file(REMOVE_RECURSE "${CURRENT_PACKAGES_DIR}/debug/include")
file(REMOVE_RECURSE "${CURRENT_PACKAGES_DIR}/include/include")
```
The header files folder were installed redundantly, and this issue can also be fixed upstream.
![image](https://github.com/user-attachments/assets/e0cff547-862c-4341-ba5c-4659a3ca8017)
![image](https://github.com/user-attachments/assets/abf594da-ec31-40ee-be80-0b16c4d3d5d7)

LilyWangLL commented:
Usage passed on `x64-windows`.

jglanz committed:
Update ports/irsdkcpp/portfile.cmake

Accept this for now, will also fix in v1.1.0 (next release)

Co-authored-by: Lily Wang <94091114+LilyWangLL@users.noreply.github.com>

jglanz committed:
Update ports/irsdkcpp/usage

@lilywangll awesome use of generator, I'll pull the namespace up in the next release

Co-authored-by: Lily Wang <94091114+LilyWangLL@users.noreply.github.com>

jglanz committed:
[irsdkcpp] Updated git-tree after commit of suggestions

LilyWangLL commented on ports/irsdkcpp/portfile.cmake:
```diff
@@ -0,0 +1,30 @@
+    REPO vrkit-platform/irsdk-cpp
+    REF "v${VERSION}"
+    SHA512 efdaafc2badcce7ba88fb943882aa301fa3c9959c8de86a99269ade05f188828c485e16f21241d558b108d48e566702e961a1c5b0bda42e523abcc783bdde5c6
+    HEAD_REF develop
```
```suggestion
    HEAD_REF master
```
The default branch of [upsteam](https://github.com/[vrkit-platform/irsdk-cpp](https://github.com/vrkit-platform/irsdk-cpp)) is master

jglanz replied:
We merge onto the develop branch, only releases hit main. Therefore in our library, `develop` is upstream (or next release/head).

LilyWangLL commented on ports/irsdkcpp/usage:
```diff
@@ -0,0 +1,4 @@
+irsdkcpp provides CMake targets:
+
+  find_package(irsdkcpp CONFIG REQUIRED)
+  target_link_libraries(main PRIVATE $<IF:$<TARGET_EXISTS:irsdkcpp::irsdkcpp_shared>,irsdkcpp::irsdkcpp_shared,irsdkcpp::irsdkcpp_static>)
```
This port also provides the `pkg-config` usage.

jglanz replied:
pkgconfig in 1.0.8 is invalid, 1.0.9 will have valid pkgconfig

jglanz committed:
[irsdkcpp] 1.0.9 update.  pkgconfig is now correct, usage updated, other cleanup

jglanz commented:
@LilyWangLL I bumped everything to the new `1.0.9` release, which contains all the fixes discussed above.  

Hope you have a great weekend & please let me know if any other changes are needed

LilyWangLL commented on ports/irsdkcpp/usage:
```diff
@@ -0,0 +1,15 @@
+
+  
+irsdkcpp provides pkg-config modules:
+# NOTE: The pkg-config modules require explicit linkage to either `irsdkcpp_shared` or `irsdkcpp_static`.  
+#   Additionally `DEBUG` builds have a `d` suffix
+  irsdkcpp
```
```suggestion
  irsdkcpp

```

LilyWangLL commented:
Please remove the comments in `usage`.

jglanz committed:
[irsdkcpp] removed comments from usage per request of @LilyWangLL
